### PR TITLE
DOC: Fix flake8 issues with whatsnew v0.18.*

### LINKED ISSUE #24233: DOC: Fix flake8 issues in doc/source/whatsnew/v0.18.*.rst
xref: #24177 

We didn't start validating the format of PEP8 and other code standards in the documentation examples until recently. We still have some files with errors, that we need to skip, and that we should fix, so we can start validating them.

**Two steps are required to replicate the issue**

1. Edit `setup.cfg` in the pandas home, and in the flake8-rst section, remove from the `exclude` list the file `doc/source/whatsnew/v0.18.0.rst` & `doc/source/whatsnew/v0.18.1.rst`
2. Remove the import block ( `from pandas import *`) from the lines (8-11) of the file `v0.18.*.rst`. 

After that, running the next command will report the errors in the file (note that syntax error usually prevent to validate other errors, and the list of errors to fix can become much longer when the syntax error is fixed (please make sure that you are using `flake8-rst` version 0.7.0 or higher):

```
$ flake8-rst doc/source/whatsnew/v0.18.*
doc/source/whatsnew/v0.18.0.rst:62:26: E203 whitespace before ':'
doc/source/whatsnew/v0.18.0.rst:62:43: E203 whitespace before ':'
doc/source/whatsnew/v0.18.0.rst:69:30: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:100:13: E999 SyntaxError: invalid syntax
doc/source/whatsnew/v0.18.0.rst:120:14: E203 whitespace before ':'
doc/source/whatsnew/v0.18.0.rst:120:24: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:121:14: E203 whitespace before ':'
doc/source/whatsnew/v0.18.0.rst:121:24: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:199:18: W605 invalid escape sequence '\d'
doc/source/whatsnew/v0.18.0.rst:214:10: W605 invalid escape sequence '\d'
doc/source/whatsnew/v0.18.0.rst:220:10: W605 invalid escape sequence '\d'
doc/source/whatsnew/v0.18.0.rst:268:31: W605 invalid escape sequence '\d'
doc/source/whatsnew/v0.18.0.rst:274:31: W605 invalid escape sequence '\d'
doc/source/whatsnew/v0.18.0.rst:287:19: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:287:23: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:287:30: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:288:19: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:288:23: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:288:30: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:292:27: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:292:31: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:292:38: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:327:50: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:327:60: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:354:28: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:354:30: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:376:20: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:376:22: E231 missing whitespace after ','
doc/source/whatsnew/v0.18.0.rst:725:83: E501 line too long (83 > 79 characters)
doc/source/whatsnew/v0.18.1.rst:51:10: F821 undefined name 'datetime'
doc/source/whatsnew/v0.18.1.rst:170:4: E124 closing bracket does not match visual indentation
doc/source/whatsnew/v0.18.1.rst:185:83: E501 line too long (80 > 79 characters)
doc/source/whatsnew/v0.18.1.rst:315:9: F821 undefined name 'DataFrame'
doc/source/whatsnew/v0.18.1.rst:315:23: E203 whitespace before ':'
doc/source/whatsnew/v0.18.1.rst:316:23: E203 whitespace before ':'
doc/source/whatsnew/v0.18.1.rst:437:92: E501 line too long (84 > 79 characters)
doc/source/whatsnew/v0.18.1.rst:457:93: E501 line too long (84 > 79 characters)
doc/source/whatsnew/v0.18.1.rst:476:29: F821 undefined name 'StringIO'
doc/source/whatsnew/v0.18.1.rst:480:29: F821 undefined name 'StringIO'
doc/source/whatsnew/v0.18.1.rst:488:29: F821 undefined name 'StringIO'
doc/source/whatsnew/v0.18.1.rst:492:29: F821 undefined name 'StringIO'
```

Once all the errors are addressed, please open a pull request with the fixes in the file, and removing the file from setup.cfg. If you need to do something that feels wrong to fix an error, please ask in a comment to this issue. Please avoid other unrelated changes, which can be addressed in a separate pull request.
cc: @datapythonista

### PR DESCRIPTION
**Pull request by JimJeon**
Fixed almost every flake8 issues in v0.18.*.rst except one issue
about tab-completion.

- [x] closes #24233 
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

I fixed almost every flake8 issues except one.
An issue about **tab-completion** and I have no idea how to fix it.

In `doc/source/whatsnew/v0.18.0.rst` there is a code like below.
```python
In [9]: r.<TAB>
r.A           r.agg         r.apply       r.count       r.exclusions  r.max         r.median      r.name        r.skew        r.sum
r.B           r.aggregate   r.corr        r.cov         r.kurt        r.mean        r.min         r.quantile    r.std         r.var
```

I've searched with `grep` in every whatsnew files but there was only one other example about this in `v0.17.0.rst`. Would you review this PR?

### CODE REVIEW

**Comment by saurav2608** [comment]
thanks @JimJeon : Please wait for a formal review from the maintainers.  I just want to point that you have to remove the block between 8 & 11 for both the files (the below chunk should be deleted). Removal should generate some more errors. 

```
.. ipython:: python
   :suppress:

   from pandas import * # noqa F401, F403
```

**Comment by datapythonista** [comment]
also, for the tab error, we use a noqa, do a `grep "<TAB>" *.rst` to see other cases

File: doc/source/whatsnew/v0.18.1.rst
Comment (datapythonista) [inline doc/source/whatsnew/v0.18.1.rst]: can you use `import datetime` instead, and use `datetime.datetime(...)` below

File: doc/source/whatsnew/v0.18.1.rst
Comment (datapythonista) [inline doc/source/whatsnew/v0.18.1.rst]: I found this indentation confusing. If the lines are too long, can we use something like:
```
dft2 = pd.DataFrame(
    np.random.randn(20, 1),
    ...
```

File: doc/source/whatsnew/v0.18.1.rst
Comment (datapythonista) [inline doc/source/whatsnew/v0.18.1.rst]:
```suggestion
.. code-block:: ipython
```

File: doc/source/whatsnew/v0.18.1.rst
Comment (datapythonista) [inline doc/source/whatsnew/v0.18.1.rst]: can you use `import io` instead, and have the import visible in the next block, where it's used.

**Review by datapythonista — COMMENTED** [review]
Looks good, added some comments, also what @saurav2608 said and the tab.

File: doc/source/whatsnew/v0.18.1.rst
Comment (JimJeon) [reply doc/source/whatsnew/v0.18.1.rst]: Sure. May I ask why you prefer that?
Is that because it can cause some namespace issues?

File: doc/source/whatsnew/v0.18.1.rst
Comment (JimJeon) [reply doc/source/whatsnew/v0.18.1.rst]: Thanks for the comment. I wasn't flexible with that. I'll fix it.

File: doc/source/whatsnew/v0.18.1.rst
Comment (datapythonista) [reply doc/source/whatsnew/v0.18.1.rst]: There are many reasons, and some are not trivial to explain.

The main one is that if I see a bare `datetime` in Python code, I need to guess whether it's the main module, or the `datetime.datetime` module. While there is no ambiguity in reading `datetime.datetime`.

Then, with a `from datetime import datetime` I load the whole `datetime` module in `sys.modules`, but I only have access in the namespace to the `datetime.datetime` submodule (with the name `datetime`). This won't let me perform operations like `imp.reload(datetime)`, as I don't have anything in the namespace pointing to the main `datetime` module.

**Comment by JimJeon** [comment]
@saurav2608 @datapythonista Thanks for the reviews!

File: doc/source/whatsnew/v0.18.1.rst
Comment (datapythonista) [inline doc/source/whatsnew/v0.18.1.rst]: can you move the import to the beginning of the next code block, and remove this block.

File: doc/source/whatsnew/v0.18.1.rst
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.rst]: I think you need a blank line here

**Comment by jreback** [comment]
some comments. ping on green.

File: doc/source/whatsnew/v0.18.1.rst
Comment (JimJeon) [reply doc/source/whatsnew/v0.18.1.rst]: Thanks.

**Review by datapythonista — APPROVED** [review]
lgtm, thanks @JimJeon

**Comment by jreback** [comment]
thanks

**Pull request merged**
